src/QsCompiler/CommandLineTool/CompilationTracker.cs

```diff
@@ -0,0 +1,281 @@
+            /// Halts the stopwatch of the compilation task and stores its duration.
+            /// </summary>
+            public void End()
+            {
+                UtcEnd = DateTime.UtcNow;
+                Watch.Stop();
+                DurationInMs = Watch.ElapsedMilliseconds;
+            }
+
+            /// <summary>
+            /// Returns whether a compilation class is in progress.
+            /// </summary>
+            public bool IsInProgress()
+            {
+                return Watch.IsRunning;
+            }
+        }
+
+        /// <summary>
+        /// Represents a node in the tree of tasks performed by the compiler.
+        /// </summary>
+        private class CompilationTaskNode
+        {
+            public readonly CompilationTask Task;
+            public readonly IDictionary<string, CompilationTaskNode> Children;
+
+            public CompilationTaskNode(CompilationTask task)
+            {
+                Task = task;
+                Children = new Dictionary<string, CompilationTaskNode>();
+            }
+        }
+
+        /// <summary>
+        /// Represents a warning type detected while tracking compilation events.
+        /// </summary>
+        private enum WarningType
+        {
+            ProcessAlreadyExists,
+            ProcessDoesNotExist,
+            ProcessAlreadyEnded
+        }
+
+        /// <summary>
+        /// Represents a warning detected while tracking compilation events.
+        /// </summary>
+        private class Warning
+        {
+            public readonly WarningType Type;
+            public readonly DateTime UtcDateTime;
+            public readonly string Key;
+
+            public Warning(WarningType type, string key)
+            {
+                UtcDateTime = DateTime.UtcNow;
+                Type = type;
+                Key = key;
+            }
+        }
+
+        // Private members.
+
+        /// <summary>
+        /// Defines a handler for a type of compilation task event.
+        /// </summary>
+        private delegate void CompilationTaskEventTypeHandler(CompilationLoader.CompilationTaskEventArgs eventArgs);
+
+        /// <summary>
+        /// Contains a handler that takes care of each type of task event.
+        /// </summary>
+        private static readonly IDictionary<CompilationLoader.CompilationTaskEventType, CompilationTaskEventTypeHandler> CompilationEventTypeHandlers = new Dictionary<CompilationLoader.CompilationTaskEventType, CompilationTaskEventTypeHandler>
+        {
+            { CompilationLoader.CompilationTaskEventType.Start, CompilationEventStartHandler },
+            { CompilationLoader.CompilationTaskEventType.End, CompilationEventEndHandler }
+        };
+
+        /// <summary>
+        /// Contains the compilation tasks tracked through the handled events.
+        /// </summary>
+        private static readonly IDictionary<string, CompilationTask> CompilationTasks = new Dictionary<string, CompilationTask>();
```

Comment:
I added a global lock to make CompilationTracker thread safe. I think it makes sense to have a global lock instead of concurrent data structures since we need a lock to protect the operation of building the tree anyway.